`fieldnames()` can throw for types such as `typeof(+)`

staticfloat opened the pull request:
Wrap its invocation here in a `try` block.  Fixes #32558

StefanKarpinski commented:
Maybe better to fix `fieldnames` instead/as well?

JeffBezanson commented:
```
julia> fieldnames(typeof(+))
()
```
?

StefanKarpinski commented:
Then what was the actual trigger here?

staticfloat commented:
I don't know how to construct the exact path that leads us to this error; you can see it by typing `typeof(+).<tab>` into the REPL.  You can also trigger it via `fieldnames(Type{typeof(+)})`, although I'm not 100% certain that's the same fundamental error.

JeffBezanson commented:
It is possible for `fieldnames` to error, but it will only do that on abstract types. It would be good to understand what's happening in the REPL. Is it supposed to reach here with an abstract type? If so, we could add a check for concreteness.

ararslan commented:
This is interesting:
```
julia> t = typeof(+)
typeof(+)

julia> t.<TAB>

abstract         isbitstype        isinlinealloc     llvm::StructType  names             size              uid
hasfreetypevars  isconcretetype    layout            mutable           ninitialized      super             zeroinit
instance         isdispatchtuple   llvm::DIType      name              parameters        types
```

ararslan commented:
> You can also trigger it via `fieldnames(Type{typeof(+)})`

The value `t` in https://github.com/JuliaLang/julia/blob/master/stdlib/REPL/src/REPLCompletions.jl#L123 becomes `Type{typeof(+)}`, so the call on the line touched by this PR indeed becomes `fieldnames(Type{typeof(+)})`.

JeffBezanson commented:
Ah, ok, so if `t` has the form `Type{T}`, we should convert it to `typeof(T)` instead. After that, we should also check `isconcretetype(t)`, and use empty fieldnames if false.

staticfloat committed:
Fix error thrown upon `typeof(+).<TAB>` at REPL

We don't want types such as `Type{typeof(+)}` to be passed to
`fieldnames()` inside of REPLCompletions.

staticfloat commented:
@JeffBezanson I've updated this PR (and rebased it on top of `master`) does this do what you described above?

staticfloat committed:
Only pull out the first type parameter if it exists

staticfloat committed:
Only sub out for the parameter if it's not already concrete

JeffBezanson commented on stdlib/REPL/src/REPLCompletions.jl:
```diff
@@ -153,11 +153,18 @@ function complete_symbol(sym, ffunc, context_module=Main)::Vector{Completion}
-                if startswith(s, name)
-                    push!(suggestions, FieldCompletion(t, field))
+            # Check for cases like Type{typeof(+)}
+            if t isa Type && !isconcretetype(t) && length(t.parameters) >= 1
```
```suggestion
            if t isa DataType && t.name === Base._TYPE_NAME
```

JeffBezanson commented on stdlib/REPL/src/REPLCompletions.jl:
```diff
@@ -153,11 +153,18 @@ function complete_symbol(sym, ffunc, context_module=Main)::Vector{Completion}
-                    push!(suggestions, FieldCompletion(t, field))
+            # Check for cases like Type{typeof(+)}
+            if t isa Type && !isconcretetype(t) && length(t.parameters) >= 1
+                t = t.parameters[1]
```
```suggestion
                t = typeof(t.parameters[1])
```

staticfloat committed:
Update stdlib/REPL/src/REPLCompletions.jl

Co-Authored-By: Jeff Bezanson <jeff.bezanson@gmail.com>

staticfloat committed:
Update stdlib/REPL/src/REPLCompletions.jl

Co-Authored-By: Jeff Bezanson <jeff.bezanson@gmail.com>

JeffBezanson commented on stdlib/REPL/test/replcompletions.jl:
```diff
@@ -1049,3 +1049,9 @@ let s = "prevind(\"θ\",1,"
+# Issue #32840
+let s = "typeof(+)."
+    c, r = test_complete_context(s)
+    @test isempty(c)
```
This won't be empty --- it should contain `fieldnames(DataType)`.

staticfloat committed:
Don't expect it to be empty

JeffBezanson commented on stdlib/REPL/test/replcompletions.jl:
```diff
@@ -103,6 +103,7 @@ let s = ""
 end
 
 let s = "using REP"
+    @show completions(s, lastindex(s))[1]
```
Should this be removed?

staticfloat committed:
remove debugging code
